drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,339 @@
+	const char *base_path,
+	const char *name,
+	int *value)
+{
+	char path[PATH_MAX];
+	int ret;
+
+	if (snprintf(path, sizeof(path), "%s/%s", base_path, name) >= PATH_MAX) {
+		errno = ENAMETOOLONG;
+		upslog_with_errno(LOG_ERR, "snprintf(%s/%s) has failed", base_path, name);
+		return -ENAMETOOLONG;
+	}
+
+	if ((ret = file_read_number(path, value)) < 0) {
+		errno = -ret;
+		upslog_with_errno(LOG_ERR, "file_read_number(%s) has failed", path);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int update_voltage(void)
+{
+	return update_intvar(ina219_base_path, "in1_input", &voltage);
+}
+
+static int update_current(void)
+{
+	return update_intvar(ina219_base_path, "curr1_input", &current);
+}
+
+void upsdrv_makevartable(void)
+{
+	addvar(VAR_VALUE, "sysfs_dir",
+			"Path to sysfs dir of hwmon (" SYSFS_HWMON_DIR ")");
+	addvar(VAR_VALUE, "lb",
+			"Low battery threshold (" STRFY(LOW_BATTERY) ")");
+	addvar(VAR_VALUE, "hb",
+			"High battery threshold (" STRFY(HIGH_BATTERY) ")");
+}
+
+void upsdrv_initups(void)
+{
+	const char *hwmon_dir = SYSFS_HWMON_DIR;
+	int ret;
+
+	if (getval("sysfs_dir"))
+		hwmon_dir = getval("sysfs_dir");
+
+	if ((ret = detect_hwmon_ina219(hwmon_dir)) < 0) {
+		errno = -ret;
+		fatal_with_errno(EXIT_FAILURE, "detect_hwmon_ina219(%s) has failed",
+				hwmon_dir);
+	}
+
+	if (getval("lb")) {
+		const int tmp = atoi(getval("lb"));
+
+		if (tmp == 0 || tmp > 100) {
+			errno = ERANGE;
+			fatal_with_errno(EXIT_FAILURE, "Invalid value from option lb: '%s'",
+					getval("lb"));
+		}
+		else {
+			low_battery = tmp;
+		}
+	}
+
+	if (getval("hb")) {
+		const int tmp = atoi(getval("hb"));
+
+		if (tmp == 0 || tmp > 100) {
+			errno = ERANGE;
+			fatal_with_errno(EXIT_FAILURE, "Invalid value from option hb: '%s'",
+					getval("hb"));
+		}
+		else {
+			high_battery = tmp;
+		}
+	}
+
+	if (low_battery > high_battery) {
+		errno = EINVAL;
+		fatal_with_errno(EXIT_FAILURE, "Invalid options: lb > hb");
+	}
+
+	upslogx(LOG_NOTICE, "lb = %u hb = %u\n", low_battery, high_battery);
+}
+
+void upsdrv_initinfo(void)
+{
+	dstate_setinfo("ups.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("ups.model", "%s", "INA219");
+	dstate_setinfo("ups.type", "%s", "ups");
+	dstate_setinfo("device.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("device.model", "%s", "INA219");
+	dstate_setinfo("device.type", "%s", "ups");
+	dstate_setinfo("device.description", "%s",
+			"Bidirectional Current/Power Monitor With I2C Interface");
+
+	dstate_setinfo("battery.charge.low", "%d", low_battery);
+	dstate_setinfo("battery.charge.high", "%d", high_battery);
+}
+
+void upsdrv_updateinfo(void)
+{
+	unsigned int charge = 0;
+	int stale = 0;
+
+	if (update_voltage() < 0)
+		stale = 1;
+
+	upsdebugx(1, "Battery voltage: %.3fV", voltage / 1000.0);
+
+	if (update_current() < 0)
+		stale = 1;
+
+	upsdebugx(1, "Battery current: %.3fA", current / 1000.0);
+
+	if (stale) {
+		dstate_datastale();
+		return;
+	}
+
+	status_init();
+
+	charge = (voltage - 3000) / 12;
```

Comment:
Note earlier comments about possibly range-checking (e.g. if accounting is truly in milli-units or can vary between systems), so whether `3000` is correct here.

In fact, this seems to assume a few numbers about the battery (high/low voltages for its dynamic range) which can depend on its chemistry (model/maker), age, etc. This should be at least commented, ideally externalized as macros or variables (setting `default.battery.voltage.high` etc. is pretty common in other drivers; note that the `default.` or `override.` prefix would be chopped by NUT core code, so your init or update methods would deal with `battery.voltage.high` whether served by HW or coming from config).

You've mentioned Li-Ion batteries, so the nominal would be around 3600mV? The 3 batteries of yours, are they in parallel or sequential connection (do their voltages sum up or average out, essentially)?

I can guess your setup dealt with `(3600 - 3000) / 12 = 600/12 = 50` (%) charge assumption most of the time when the graph is almost level? Allegedly, starting points for some chemistries can be over 4V (per https://www.jackery.com/blogs/knowledge/lithium-ion-battery-voltage-chart for example) with a quick drop to ~3.5V when discharging, so that formula could be showing close to 100% when charged.

Reply:
> You've mentioned Li-Ion batteries, so the nominal would be around 3600mV? The 3 batteries of yours, are they in parallel or sequential connection (do their voltages sum up or average out, essentially)?

I use 3x SONY VTC6 - 18650 Battery - 30A (e.g. here https://www.cigareta-shop.eu/sony-vtc6-18650-battery-30a/). Parallel setup, thus their capacity should sum up and voltage is the same. Yes, 3.6 V is the nominal voltage.

> I can guess your setup dealt with (3600 - 3000) / 12 = 600/12 = 50 (%) charge assumption most of the time when the graph is almost level? Allegedly, starting points for some chemistries can be over 4V (per https://www.jackery.com/blogs/knowledge/lithium-ion-battery-voltage-chart for example) with a quick drop to ~3.5V when discharging, so that formula could be showing close to 100% when charged.

The nominal voltage of the 18650 battery is 3,6 V. The fully charged voltage is 4,25 V. Thus, according to the section "What voltage should a lithium battery read?", it is depleted at 2,8-3 V. This is exactly what the code does:

* _(4.25 - 3) / 1.25_ - the actual formula, it should give quite exact charge in percent,
* _(4250 - 3000) / 12_ - the code is a bit shifted (miliV) and we do integer math (counting in 0..100 not in 0.01..1), thus it is not accurate and gives more than 100 % on fully charged

And empirically confirmed by the plot in this PR.

> Note earlier comments about possibly range-checking (e.g. if accounting is truly in milli-units or can vary between systems), so whether 3000 is correct here.

I believe it is correct.

>
> In fact, this seems to assume a few numbers about the battery (high/low voltages for its dynamic range) which can depend on its chemistry (model/maker), age, etc. This should be at least commented, ideally externalized as macros or variables (setting default.battery.voltage.high etc. is pretty common in other drivers; note that the default. or override. prefix would be chopped by NUT core code, so your init or update methods would deal with battery.voltage.high whether served by HW or coming from config).

OK, I am bit lost in those nut variables and their exact semantics and where they come from. Would it be ok to set:

* default.battery.voltage.high = 4250
* default.battery.voltage.low = 3000

and make somehow possible to override those values? Can you point me to a driver where do you think it is close to this situation?

Reply:
Thanks for the maby other replies abd clarifications. As for battery voltage high/low, recent-ish examples on my mind are in blazer_usb and nutdrv_qx enhancements to "guesstimation".

Typically there are static variables in the driver with default values for low/high(/nominal), and an e.g. `getval("battery.voltage.high")` would yield either NULL or user's `default` or `override` setting (same in this case). Typically you'd check for this in initups method (like your already checked `lb` etc. settings) to set those static variables for each value.

Reply:
I inspired a bit and reworked the `charge` computation. It is now more generic and useful with other battry nominal voltages as well. What do you think?

Reply:
Logic looks good, but I feel integer data types can be a problem. Commented in more detail separately.